Dev sequencing

goetzrrGit opened the pull request:
* **Tickets addressed:** Clipper's need for a new sequence editor
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description

Frontend PR: https://github.com/NASA-AMMOS/aerie-ui/pull/1275

Big push to merge the last month of work back into Core Aerie. This is merged with the database migration from 2.8. 

### High level overview of the changes:

- Modify User Sequence table to remove eDSL and include `SeqN.txt`, `SeqJSON`, and `ParcelID` https://github.com/NASA-AMMOS/aerie-ui/issues/1238
- Added Parcels containers to sequencing expansion and sequence authoring. Parcels includes the `command dictionary`, `channel dictionary`, and `parameter dictionary`, and `adaptation file` - https://github.com/NASA-AMMOS/aerie/issues/1439
- Modified the e2e test to use `Parcels` instead of `Command Dictionary`
- First pass at removing unused code and endpoints.


## Verification
Manually tested the whole expansion process. Sequencing server e2e test pass

## Next Step:
Update the test: https://github.com/NASA-AMMOS/aerie/issues/1448

cohansen committed:
First pass at the new sequence editor tables

cohansen committed:
Updated aerie-ampcs to 1.0.2 so we get hw and fsw command types on dictionary upload

cohansen committed:
Added loops and conditionals to the banananation command dictionary

joswig committed:
Sample Parameter dictionary

cohansen committed:
Changed the command dictionary table to support channel and parameter dictionaries

cohansen committed:
First pass at adding parcels

cohansen committed:
Changed user sequence command dict field to point to a parcel_id instead

cohansen committed:
Added support for a single parameter dictionary

joswig committed:
update to ampcs 1.0.4

goetzrrGit committed:
Update aerie-apmcs to 1.0.4

cohansen committed:
Added support for multiple parameter dictionaries

cohansen committed:
Added channel dictionary tables and hasura metadata

cohansen committed:
Fixed command_types_typescript_path being removed while rebasing

joswig committed:
update aerie-ampcs to 1.0.5

goetzrrGit committed:
Add seq_json column to the UserSequence Table

cohansen committed:
Added fields to support dictionary caching

goetzrrGit committed:
Update sql tables:

* Dictionary tables all match and included 'path'
* Remove edsl_string from 'expanded_sequence' table
* Included 'parcel_id' in 'expansion_rule', 'expansion_set' and 'expansion_set_to_rule'
* change the order of 'init.sql' so table were built in the right order

goetzrrGit committed:
Remove/tweak relations in hasura

goetzrrGit committed:
Updated the hasura action

* Now we can upload multiple dictionary via the action

goetzrrGit committed:
Ignore just moving files around

goetzrrGit committed:
Create a ParcelBatchLoader

goetzrrGit committed:
Update background transpiler to use Parcels

goetzrrGit committed:
Allow for multiplie dictionary upload

* backend will parse the correct dictionary

goetzrrGit committed:
Add Channel and Parameter type interface

* This is passed to the frontend for monico

goetzrrGit committed:
Expansion uses Channel and Parmeters Dictionaries

*  Update batch loader to grab all the information expansion needs
* Worker has been updated to trnaspile the new dictionary types and expand

goetzrrGit committed:
Update import paths

goetzrrGit committed:
Massive e2e Test Update

goetzrrGit committed:
Remove unused endpoints

* We are dropping eDSL from user sequences

goetzrrGit committed:
remove old references to database/hasura

* Legacy database tables and hasura metadata files the were moved in the 2.6 migration

goetzrrGit committed:
Sequence DB Migration merge

goetzrrGit committed:
Lost file found for testing

goetzrrGit committed:
Added Parcels to db-e2e sequencing testing

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/channel_dictionary.sql:
```diff
@@ -0,0 +1,36 @@
+create table sequencing.channel_dictionary (
+  id integer generated always as identity,
+
+  path text not null,
```
Major: `path` is a reserved word in PSQL and should _not_ be used as a column name. `filepath` or `json_path` are both not reserved and more descriptive alternates.

This issue occurs in `command_dictionary` and `parameter_dictionary` as well.

goetzrrGit replied:
I updated it to `dictionary_path` if that works with you?

Mythicaeda replied:
That works for me!

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/parcel.sql:
```diff
@@ -0,0 +1,42 @@
+);
+
+comment on table sequencing.parcel is e''
+  'A bundled containing dictionaries and an adaptation file.';
```
We use the phrase `adaptation file` to refer to the model `JAR` at other places in our system. Adding the word `sequencing` would be sufficient to avoid future confusion (as in `a sequencing adaptation file`). This applies to L37 as well.

goetzrrGit replied:
Good point. I made the change.

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/parcel.sql:
```diff
@@ -0,0 +1,42 @@
+  created_at timestamptz not null default now(),
+  updated_at timestamptz not null default now(),
+
+  owner text,
```
Missing fkey to `users` table

goetzrrGit replied:
Oh noooo. I added it :)

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/parcel.sql:
```diff
@@ -0,0 +1,42 @@
+
+  name text not null,
+
+  channel_dictionary_id integer default null,
+  command_dictionary_id integer not null,
```
Nonblocking, nitpick: reverse the order of these two columns

goetzrrGit replied:
I agree and made the change

Mythicaeda committed:
Update PGCMP workflow

Pull the `env` mapping to the top of the file rather than as-needed.

Explicitly install the PSQL client for PG16. We were formerly installin the PSQL client for PG14

Mythicaeda committed:
Update PGCMP workflow

- Pull the `env` mapping to the top of the file rather than as-needed.
- Explicitly install the PSQL client for PG16. We were formerly installing the PSQL client for PG14
- Use NASA-AMMOS fork of pgcmp with PG16 bug fix
- Fix reference to PG14.1

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/expanded_sequences.sql:
```diff
@@ -5,16 +5,17 @@ create table sequencing.expanded_sequences (
   seq_id text not null,
   simulation_dataset_id int not null,
   expanded_sequence jsonb not null,
-  edsl_string text not null,
```
What happened to the contents of column?

goetzrrGit replied:
It will be dropped. We will not be using this column moving forward. Sequences that have been expanded will only store the seqJSON. We will be converting the seqJSON to SeqN.txt in the editor.

Mythicaeda replied:
With regards to the `down.sql`, is this data recoverable? If not, we should call it out as one of the breaking changes.

goetzrrGit replied:
Sadly the data will not be recoverable. Is this something we let @dandelany know about? I don't really know the formal way of reporting this.

Mythicaeda replied:
I think any data we're going to drop should be mentioned to @dandelany, as he needs to include it in the Release Notes.

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/expansion_rule.sql:
```diff
@@ -5,7 +5,7 @@ create table sequencing.expansion_rule (
   expansion_logic text not null,
 
-  authoring_command_dict_id integer,
+  parcel_id integer not null,
```
Because `parcels` must be associated with a command dict, this change is implicitly saying that `expansion_rules` must now be associated with a `command_dict`. Is this intentional?

goetzrrGit replied:
I am sorry, I don't follow the question. A `Parcel` is required for each `Exapansion Rule`. Each `Parcel` will contain a `command dictionary`, optional `channel dictionary`, optional` parameter dictionary`, and an optional `Seq adaptation JavaScript` file. 

Within the UI, you'll only be able to select Parcels. We'll then automatically extract the necessary data from the chosen Parcel.

Mythicaeda replied:
So currently, `expansion_rule`s could have a `null` `command_dictionary_id`, meaning they were not required to be associated with a command dictionary. In this PR, `expansion_rules`s _must_ be associated with a parcel, which _must_ be associated with a `command_dictionary_id`, so `expansion_rule`s _must_ be associated with a command dictionary (through the `rule`'s parcel).

Additionally, the case of "`expansion_rule` with `null command_dict_id`" will likely cause issues with your migration. The easiest way to solve this would be to drop any `expansion_rule`s where `authoring_command_dict_id is null` during the data migration for this table.

goetzrrGit replied:
Got it, I added this to the `up.sql`  in the section:

```
--------------------------------------
---- Modify the Expansion Rule Table
--------------------------------------
```

add this
```
--- drop rule with no command dictionary
delete from sequencing.expansion_rule er
where authoring_command_dict_id is null;
```

Mythicaeda replied:
I'll resolve this once I've had a chance to review the `up.sql`, but that change looks sufficient.

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/parcel_to_parameter_dictionary.sql:
```diff
@@ -0,0 +1,19 @@
+create table sequencing.parcel_to_parameter_dictionary (
+  id integer generated always as identity,
```
This `id` column needs to be removed. The actual pkey of this table is `(parcel_id, param_dict_id)`, as this is an association table. Without this change, one parcel can have multiple references to the same param dict.

goetzrrGit replied:
Thanks for the suggestion. I removed the `id` and the `parcel_to_parameter_dictionary_synthetic_key` and modified the UI with these changes. Manually testing passed

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/sequence_adaptation.sql:
```diff
@@ -0,0 +1,22 @@
+
+  created_at timestamptz not null default now(),
+  updated_at timestamptz not null default now(),
+  updated_by text,
```
Non-blocking: would an `owner` field make sense for this table? The `owner` would be set to the initial creator of the adaptation if the field is desired.

goetzrrGit replied:
I will add owner to the table :)

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/user_sequence.sql:
```diff
@@ -1,24 +1,24 @@
   created_at timestamptz not null default now(),
   definition text not null,
+  seq_json jsonb,
   id integer generated always as identity,
```
Nonblocking, readability: Since `id` is the pkey for this table, can it be moved to the top (L2)?

goetzrrGit replied:
I am all for readability!!! I made the change.

Mythicaeda commented on deployment/hasura/metadata/databases/tables/tables.yaml:
```diff
@@ -154,6 +154,11 @@
 - "!include sequencing/sequence.yaml"
 - "!include sequencing/sequence_to_simulated_activity.yaml"
 - "!include sequencing/user_sequence.yaml"
+- "!include sequencing/channel_dictionary.yaml"
+- "!include sequencing/parameter_dictionary.yaml"
+- "!include sequencing/parcel.yaml"
+- "!include sequencing/parcel_to_parameter_dictionary.yaml"
+- "!include sequencing/sequence_adaptation.yaml"
```
Nonblocking: The order of these `include` statements is intended to be in alphabetical order.

goetzrrGit replied:
No problem. Made the change as well.

Mythicaeda commented on deployment/hasura/metadata/actions.graphql:
```diff
@@ -234,13 +211,14 @@ type TypescriptFile {
+type DictionaryResponse {
   id: Int!
-  command_types_typescript_path: String!
+  path: String!
```
Nonblocking: This `path` shouldn't be problematic, but I'm flagging this in case you want to rename it to whatever you rename the DB columns to.

goetzrrGit replied:
I renamed this as well and updated the UI.

Mythicaeda commented on deployment/hasura/metadata/actions.graphql:
```diff
@@ -234,13 +211,14 @@ type TypescriptFile {
+  path: String!
   mission: String!
   version: String!
   parsed_json: CommandDictionary!
```
Is this still always a CommandDictionary? If not, the type may also need to be renamed. If so, the field may need to be renamed.

goetzrrGit replied:
There is going to be `Channel` and `Parameter` dictionaries as well so I named it the `dictionary_path`

Mythicaeda replied:
I think GH flagged the wrong line. I was referring to L219: `parsed_json: CommandDictionary!`, specifically about the name of the type `CommandDictionary`

goetzrrGit replied:
Got it, yeah I change the response to be

`parsed_json: Dictionary!`

Mythicaeda commented on deployment/hasura/metadata/databases/tables/sequencing/parcel.yaml:
Nonblocking: the preferred order of these permissions is `select`, `insert`, `update`, `delete`.

goetzrrGit replied:
I didn't know this, I made the change.

Mythicaeda replied:
No worries! I need to write a Style Guide for the metadata so that this kind of thing is accessible.

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/parcel.sql:
```diff
@@ -0,0 +1,42 @@
+  created_at timestamptz not null default now(),
+  updated_at timestamptz not null default now(),
+
+  owner text,
```
Nonblocking: Since `parcels` can be updated, consider adding an `updated_by` field.

goetzrrGit replied:
I will add `update_by`

Mythicaeda commented on deployment/hasura/metadata/databases/tables/sequencing/parcel.yaml:
```diff
@@ -0,0 +1,60 @@
+  - role: aerie_admin
+    permission:
+      columns: "*"
+      check: {}
```
Minor: `set` field is missing. It goes after `check`

```suggestion
      check: {}
      set: 
          owner: "x-hasura-user-id"
```

goetzrrGit replied:
I updated the `insert_permissions:` to have the `set`

Mythicaeda commented on deployment/hasura/metadata/databases/tables/sequencing/parcel.yaml:
```diff
@@ -0,0 +1,60 @@
+update_permissions:
+  - role: aerie_admin
+    permission:
+      columns: [channel_dictionary_id, command_dictionary_id, name, sequence_adaptation_id]
```
In general, we permit `aerie_admin` to change `owner` fields. A simple use case is if the former owner leaves the project, the `owner` can be updated to who is now in charge of the thing (in this case, the parcel).

Mythicaeda commented on deployment/hasura/metadata/databases/tables/sequencing/sequence_adaptation.yaml:
```diff
@@ -0,0 +1,37 @@
+  - role: aerie_admin
+    permission:
+      columns: [adaptation]
+      check: {}
```
Minor: missing `set` here to initialize the `updated_by` column:
```suggestion
insert_permissions:
  - role: aerie_admin
    permission:
      columns: [adaptation]
      check: {}
      set:
          updated_by: "x-hasura-user-id"
```

goetzrrGit replied:
I updated and reordered the permissions to be `select, insert, update, delete`

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/user_sequence.sql:
```diff
@@ -1,24 +1,24 @@
-  authoring_command_dict_id integer not null,
   created_at timestamptz not null default now(),
   definition text not null,
+  seq_json jsonb,
```
Minor: a `comment on` for this new column is missing

goetzrrGit replied:
I added the comment for seq_json

Mythicaeda commented on deployment/hasura/metadata/databases/tables/sequencing/channel_dictionary.yaml:
I noticed that the `channel_dictionary` table contains an `updated_at` field, but that no users can update it. Is this table updated by the sequencing server? If not, consider either adding `update` permissions or removing the `updated_at` column.

This concern also applies to `command_dictionary` and `parameter_dictionary`.

goetzrrGit replied:
The `channel`, `command`, and `parameter` dictionaries are only updated by the sequencing server. We want to keep things on the backend for sequencing as much as possible.

Mythicaeda commented on deployment/hasura/metadata/databases/tables/sequencing/expansion_rule.yaml:
Nonblocking: would an object relationship from `expansion_rule` to `parcel` be of any value?

goetzrrGit replied:
I am not too sure. I can pull the information needed for expansion via the `expansion_set`.

Mythicaeda commented on sequencing-server/test/testUtils/Parcel.ts:
```diff
@@ -0,0 +1,49 @@
+export async function removeParcel(graphqlClient: GraphQLClient, parcelId: number): Promise<void> {
+  return graphqlClient.request(
+    gql`
+      mutation PutParcel($parcelId: Int!) {
```
Nonblocking: name suggestion
```suggestion
      mutation RemoveParcel($parcelId: Int!) {
```

goetzrrGit replied:
Copy/pasta :P I renamed it.

Mythicaeda commented on sequencing-server/test/testUtils/CommandDictionary.ts:
```diff
@@ -10,25 +10,25 @@ export const commandDictionaryString = fs.readFileSync(
     gql`
-      mutation PutCommandDictionary($dictionary: String!) {
-        uploadDictionary(dictionary: $dictionary) {
+      mutation PutCommandDictionary($dictionary: String!, $type: String!) {
+        uploadDictionary(dictionary: $dictionary, type: $type) {
```
`type` parameter in this mutation is unused

goetzrrGit replied:
If you go to line 41 you can see I am passing in `'COMMAND'` for the type.

Mythicaeda replied:
Ah, no, this was my bad. I was referring to the `$type` parameter passed in, but for some reason it didn't click for me that it was being used in the `uploadDictionary` action.

Mythicaeda commented on sequencing-server/src/lib/batchLoaders/parcelBatchLoader.ts:
```diff
@@ -0,0 +1,59 @@
+        query GetParcels($parcelIds: [Int!]!) {
+          parcel(where: { id: { _in: $parcelIds } }) {
+            id
+            parameter_dictionary(where: { parcel_id: { _in: $parcelIds } }) {
```
This `where` is unnecessary: `parameter_dictionary` will fetch the dictionaries associated with the specific parcel.

goetzrrGit replied:
Thank you for the suggestion, I made the change, and my manual test passed.

Mythicaeda commented on deployment/hasura/metadata/databases/tables/sequencing/parcel.yaml:
```diff
@@ -0,0 +1,60 @@
+    using:
+      foreign_key_constraint_on: sequence_adaptation_id
+array_relationships:
+  - name: parameter_dictionary
```
Array relationships should have "plural" names, i.e `parameter_dictionaries`.

goetzrrGit replied:
Thank you for caught that, I made it plural.

Mythicaeda commented on sequencing-server/src/lib/batchLoaders/expansionSetBatchLoader.ts:
```diff
@@ -62,15 +58,53 @@ export const expansionSetBatchLoader: BatchLoader<
+        }),
+      );
+
+      const paramerterResults: {
```
Minor: typo
```suggestion
      const parameterResults: {
```

goetzrrGit replied:
sigh...those typos always get me. I fixed it.

Mythicaeda commented on sequencing-server/src/app.ts:
```diff
@@ -134,27 +153,53 @@ app.get('/health', (_: Request, res: Response) => {
+  dictionaryPath = await processDictionary(parsedDictionary, type);
+  logger.info(`lib generated - path: ${dictionaryPath}`);
+
+  let db_name = 'command_dictionary';
```
Minor, nonblocking: `table_name` or `db_table_name` are more accurate options.

goetzrrGit replied:
Already had the file (app.ts) open so I made the change to `db_table_name` :P

Mythicaeda commented on package-lock.json:
Could you clarify why a project-wide `package-lock.json` is being added? Especially as this only appears to add `playwright`, which isn't a node package used in the backend anymore.

goetzrrGit replied:
Oh my gosh!!!! I didn't even notice this. Let me talk with @cohansen as the commit was when I was on vacation when he added this. I am pretty sure it is a mistake.

goetzrrGit replied:
I confirmed with Cody that it is a mistake. I removed it.

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/sequence_adaptation.sql:
```diff
@@ -0,0 +1,22 @@
+
+  created_at timestamptz not null default now(),
+  updated_at timestamptz not null default now(),
+  updated_by text,
```
Minor: missing fkey to `users` table.

goetzrrGit committed:
Server side improvements from the PR comments.

goetzrrGit committed:
hasura metadata improvements from the PR comments.

goetzrrGit committed:
database improvements from the PR comments.

goetzrrGit committed:
db migration script

goetzrrGit committed:
Delete this file

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/parcel_to_parameter_dictionary.sql:
```diff
@@ -0,0 +1,15 @@
+create table sequencing.parcel_to_parameter_dictionary (
+
```
Nonblocking: extra space :-P

goetzrrGit replied:
I took care of that extra space 😈

Mythicaeda commented on deployment/hasura/migrations/Aerie/3_sequence_parcel/up.sql:
```diff
@@ -0,0 +1,370 @@
+alter table sequencing.expanded_sequences
+  drop column edsl_string;
+
+call migrations.mark_migration_applied('3');
```
This statement is needed at the bottom of the `applied_migrations.sql` file as well

goetzrrGit replied:
Got it, I just added it.

Mythicaeda commented on deployment/hasura/migrations/Aerie/3_sequence_parcel/down.sql:
```diff
@@ -0,0 +1,207 @@
+  drop constraint parcel_channel_dictionary_id_fkey;
+drop table sequencing.channel_dictionary;
+
+--------------------------------------
+---- Drop the Parcel Table
+--------------------------------------
+drop table sequencing.parcel;
```
This block should be moved to above L183. Most importantly, this would make the `down` and the `up` consistent, but this would also avoid the need to alter this table on L186 and L193

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/parcel.sql:
```diff
@@ -0,0 +1,46 @@
+  constraint parcel_synthetic_key
+    primary key (id),
+
+  foreign key (channel_dictionary_id)
+    references sequencing.channel_dictionary (id),
+  foreign key (command_dictionary_id)
+    references sequencing.command_dictionary (id),
+  foreign key (sequence_adaptation_id)
+    references sequencing.sequence_adaptation (id),
```
All of these are missing `on update` and `on delete`

`command_dictionary` should have a `cascade` on delete, the other two should have `set null`

goetzrrGit commented:
TODO: Make a ticket for db test : DONE

goetzrrGit committed:
Last bit of changes.

goetzrrGit commented:
@Mythicaeda here is a new ticket for updating the test.

https://github.com/NASA-AMMOS/aerie/issues/1448

Mythicaeda commented on deployment/postgres-init-db/sql/tables/sequencing/parcel.sql:
```diff
@@ -0,0 +1,54 @@
+  created_at timestamptz not null default now(),
+  owner text,
+  updated_at timestamptz not null default now(),
+  update_by text,
```
Typo: 
```suggestion
  updated_by text,
```

goetzrrGit replied:
whoops. I also fixed it on the `up.sql`

Mythicaeda commented on deployment/hasura/metadata/databases/tables/sequencing/parcel.yaml:
```diff
@@ -0,0 +1,62 @@
+      columns: "*"
+      filter: {}
+      allow_aggregations: true
+insert_permissions:
+  - role: aerie_admin
+    permission:
+      columns: "*"
+      check: {}
+      set:
+        owner: "x-hasura-user-id"
+update_permissions:
+  - role: aerie_admin
+    permission:
+      columns: [channel_dictionary_id, command_dictionary_id, name, sequence_adaptation_id]
+      filter: {}
```
Missing 
```yaml
set:
   updated_by: "x-hasura-user-id"
```
on both `update` and `insert` permissions.

goetzrrGit replied:
`Insert` has this already, should I add this to `delete`

Mythicaeda replied:
No

Mythicaeda replied:
The change you is incorrectly setting `owner` instead of `updated_by` on update. The final yaml for the insert and update fields should look like this:

```yaml
insert_permissions:
  - role: aerie_admin
    permission:
      columns: "*"
      check: {}
      set:
        owner: "x-hasura-user-id"
        updated_by: "x-hasura-user-id"
update_permissions:
  - role: aerie_admin
    permission:
      columns: [channel_dictionary_id, command_dictionary_id, name, sequence_adaptation_id, owner]
      filter: {}
      set:
        updated_by: "x-hasura-user-id"
```

Note that in addition to updating the `set` fields of the `insert` and `update` permissions, this also adds the `owner` column to the set of updatable columns, as that column was previously missing.

goetzrrGit replied:
ohhhhh, Sorry about that, and thank you for explaining it to me in more detail. I made the change.

Mythicaeda commented on deployment/hasura/migrations/Aerie/3_sequence_parcel/up.sql:
```diff
@@ -0,0 +1,378 @@
+execute function util_functions.set_updated_at();
+
+--- populate table with data
+insert into sequencing.parcel (name, channel_dictionary_id, command_dictionary_id, sequence_adaptation_id, created_at, updated_at, owner)
+select
+    'Parcel ' || cd.id as name,
+    null,
+    cd.id,
+    null,
+    now(),
+    now(),
+    'Aerie Legacy'
+from sequencing.command_dictionary cd;
```
Nonblocking: this insert can be simplified. 
Blocking: "updated_by" is missing from this insert.
```sql
insert into sequencing.parcel (name, command_dictionary_id, owner, updated_by)
select
    'Parcel ' || cd.id as name,
    cd.id,
    'Aerie Legacy',
    'Aerie Legacy'
from sequencing.command_dictionary cd;
```

goetzrrGit replied:
Thank you for the modified SQL statement. I updated the `up.sql `

Mythicaeda commented on deployment/hasura/migrations/Aerie/3_sequence_parcel/down.sql:
```diff
@@ -0,0 +1,203 @@
+  'Command dictionary the user sequence was created with.';
+
+--- Data Migration
+update sequencing.user_sequence
+set authoring_command_dict_id = p.command_dictionary_id
+from sequencing.user_sequence us
+       inner join sequencing.parcel p on us.parcel_id = p.id;
```
This statement can be reworked to remove the `join` similar to its `up` counterpart:

```sql
update sequencing.user_sequence us
set authoring_command_dict_id = p.command_dictionary_id
from sequencing.parcel p 
where us.parcel_id = p.id;
```

goetzrrGit replied:
omg I am so sorry, I replaced it in the `up.sql` but forgot to replace the `down`. Thank you for caughting that.

goetzrrGit committed:
fixed the wrong changes I did in the hasura ymal.

Mythicaeda commented on deployment/hasura/migrations/Aerie/3_sequence_parcel/down.sql:
```diff
@@ -0,0 +1,203 @@
+  'The ID of a command dictionary';
+
+--- Data Migration
+update sequencing.expansion_set
+set command_dict_id = p.command_dictionary_id
+from sequencing.expansion_set es
+       inner join sequencing.parcel p on es.parcel_id = p.id;
```
This `update` can be reworked to avoid the join:
```sql
update sequencing.expansion_set es
set command_dict_id = p.command_dictionary_id
from sequencing.parcel p 
where es.parcel_id = p.id;
```

Mythicaeda commented on deployment/hasura/migrations/Aerie/3_sequence_parcel/down.sql:
```diff
@@ -0,0 +1,203 @@
+comment on column sequencing.expansion_rule.authoring_command_dict_id is e''
+  'The id of the command dictionary to be used for authoring of this expansion.';
+
+--- Data Migration
+update sequencing.expansion_rule
+set authoring_command_dict_id = p.command_dictionary_id
+from sequencing.expansion_rule er
+       inner join sequencing.parcel p on er.parcel_id = p.id;
```
This `update` can be reworked to avoid the join:
```sql
update sequencing.expansion_rule er
set authoring_command_dict_id = p.command_dictionary_id
from sequencing.parcel p 
where er.parcel_id = p.id;
```

Mythicaeda commented on deployment/hasura/migrations/Aerie/3_sequence_parcel/down.sql:
```diff
@@ -0,0 +1,203 @@
+---- Modify the Command Dictionary Table
+------------------------------------------
+
+alter table sequencing.command_dictionary
+  add column command_types_typescript_path text;
+
+--- Data Migration
+update sequencing.command_dictionary
+set command_types_typescript_path = dictionary_path
+where command_types_typescript_path is null;
+
+
+alter table sequencing.command_dictionary
+  alter column command_types_typescript_path set not null;
+
+---
+
+drop trigger set_timestamp on sequencing.command_dictionary;
+
+comment on column sequencing.command_dictionary.command_types_typescript_path is e''
+  'The location of command dictionary types (.ts) on the filesystem';
+
+alter table sequencing.command_dictionary
+  drop column updated_at,
+  drop column dictionary_path;
```
This part of the migration can be simplified, since this is just a column rename and drop:

```sql
------------------------------------------
---- Modify the Command Dictionary Table
------------------------------------------
drop trigger set_timestamp on sequencing.command_dictionary;

alter table sequencing.command_dictionary
  rename column dictionary_path to command_types_typescript_path;

alter table sequencing.command_dictionary
  drop column updated_at;
```

goetzrrGit replied:
with this change do I have to create a comment still:

```
comment on column sequencing.command_dictionary.command_types_typescript_path is e''
  'The location of command dictionary types (.ts) on the filesystem';
```

Mythicaeda replied:
Nope! The text is the same, so it'll get auto-adjusted to the new column name.

goetzrrGit replied:
Awesome

Mythicaeda commented on deployment/hasura/migrations/Aerie/3_sequence_parcel/up.sql:
```diff
@@ -0,0 +1,375 @@
+  drop column authoring_command_dict_id;
+
+
+---------------------------------------
+--- Modify the Command Dictionary Table
+---------------------------------------
+
+alter table sequencing.command_dictionary
+  add column dictionary_path text,
+  add column updated_at timestamptz not null default now();
+
+comment on column sequencing.command_dictionary.dictionary_path is e''
+  'The location of command dictionary types (.ts) on the filesystem';
+
+create trigger set_timestamp
+  before update on sequencing.command_dictionary
+  for each row
+execute function util_functions.set_updated_at();
+
+--- Data Migration
+update sequencing.command_dictionary
+set dictionary_path = command_types_typescript_path
+where dictionary_path is null;
+
+alter table sequencing.command_dictionary
+  alter column dictionary_path set not null;
+
+---
+
+alter table sequencing.command_dictionary
+  drop column command_types_typescript_path;
```
This part of the migration can be simplified since it is just a column rename and add:
```sql
---------------------------------------
--- Modify the Command Dictionary Table
---------------------------------------
alter table sequencing.command_dictionary
  add column updated_at timestamptz not null default now();

alter table sequencing.command_dictionary
 rename column command_types_typescript_path to dictionary_path;

create trigger set_timestamp
  before update on sequencing.command_dictionary
  for each row
execute function util_functions.set_updated_at();
```

goetzrrGit replied:
similar question about the `comment` do I still need to create one if we do a column rename?

goetzrrGit replied:
awesome I made the change

Mythicaeda commented on deployment/hasura/migrations/Aerie/3_sequence_parcel/down.sql:
```diff
@@ -0,0 +1,203 @@
+
+------------------------------------
+--- Modify Expanded Sequences Table
+------------------------------------
+alter table sequencing.expanded_sequences
+  add column edsl_string text;
+
+update sequencing.expanded_sequences
+set edsl_string = '{}'
+where edsl_string is null;
+
+alter table sequencing.expanded_sequences
+  alter column edsl_string set not null;
```
Nonblocking: this part of the migration could be simplified slightly by temporarily giving the column a `default` value:
```sql
------------------------------------
--- Modify Expanded Sequences Table
------------------------------------
alter table sequencing.expanded_sequences
  add column edsl_string text not null default '{}';

alter table sequencing.expanded_sequences
  alter column edsl_string drop default;
```

goetzrrGit replied:
Cool, I made the change.

Mythicaeda reviewed (commented):
Just some small changes to the migrations!

goetzrrGit committed:
Better migration queries for the up/down.sql

Mythicaeda reviewed (approved):
Awesome work! LGTM!
